Python/optimizer_bytecodes.c

```diff
@@ -167,24 +167,81 @@ dummy_func(void) {
     }
 
     op(_BINARY_OP, (left, right -- res)) {
-        PyTypeObject *ltype = sym_get_type(left);
-        PyTypeObject *rtype = sym_get_type(right);
-        if (ltype != NULL && (ltype == &PyLong_Type || ltype == &PyFloat_Type) &&
-            rtype != NULL && (rtype == &PyLong_Type || rtype == &PyFloat_Type))
-        {
-            if (oparg != NB_TRUE_DIVIDE && oparg != NB_INPLACE_TRUE_DIVIDE &&
-                ltype == &PyLong_Type && rtype == &PyLong_Type) {
-                /* If both inputs are ints and the op is not division the result is an int */
-                res = sym_new_type(ctx, &PyLong_Type);
+        bool lhs_int = sym_matches_type(left, &PyLong_Type);
+        bool rhs_int = sym_matches_type(right, &PyLong_Type);
+        bool lhs_float = sym_matches_type(left, &PyFloat_Type);
+        bool rhs_float = sym_matches_type(right, &PyFloat_Type);
+        if ((!lhs_int && !lhs_float) || (!rhs_int && !rhs_float)) {
+            res = sym_new_unknown(ctx);
+            goto binary_op_done;
+        }
+        if (oparg == NB_POWER || oparg == NB_INPLACE_POWER) {
+            // This one's fun: the *type* of the result depends on the *values*
```

Comment:
I'm not convinced that this is worth the complexity for anything other than integer constant right hand sides.
It the rhs is an int constants, then the result is the type of the lhs (for ints and floats) otherwise, "not known".
Calculating any other result is so slow that the cost of the guard we could potentially remove is negligible.

Reply:
Doesn't a negative int rhs produce a float?

```
>>> 3 ** -1
0.3333333333333333
```

Your point may still stand that it's too expensive, but unfortunately the assumption that if rhs is an int we can determine the result without looking at the value doesn't hold.